posix: implement `uname`

ycsin opened the pull request:
Implementation for posix api: `uname`

```
*** Booting Zephyr OS build zephyr-v3.4.0-633-gc38dd60da451 ***

Printing everything in utsname...
sysname[7]: Zephyr
nodename[7]: zephyr
release[9]: 3.4.99
version[61]: zephyr-v3.4.0-633-gc38dd60da451 Jul  6 2023 10:29:50
machine[8]: riscv64


uart:~$ uname -a
Zephyr zephyr 3.4.99 zephyr-v3.4.0-633-gc38dd60da451 Jul  6 2023 10:29:50 riscv64 qemu_riscv64
```

MacOS as a reference:
```
(.venv) ycsin-mba:zephyrproject ycsin$ uname -a
Darwin ycsin-mba 22.5.0 Darwin Kernel Version 22.5.0: Mon Apr 24 20:53:44 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T8103 arm64
```

fixes #59924

- [x] update `doc/services/portability/posix.rst`
- [x] add to the testsuite
- [x] [size_t](https://stackoverflow.com/a/1409631)

ycsin committed:
posix: implement uname

Add implementation for posix uname.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,72 @@
+#ifndef ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+#define ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+
+#include <errno.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/net/hostname.h>
+#include "version.h"
```
Just a minor nit - I can't remember which RFC it was but we should preemptively adopt @gmarull 's suggestion for canonical header ordering.

```suggestion
#include "version.h"

#include <errno.h>

#include <zephyr/net/hostname.h>
#include <zephyr/toolchain/common.h>
```

cfriedt commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,72 @@
+/* machine */
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
+
+struct utsname {
```
Minor comments - it looks as though the array length should be unspecified according to the spec, but if that would cause errors, then no problem.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_utsname.h.html

cfriedt commented on lib/posix/utsname.c:
```diff
@@ -0,0 +1,24 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/posix/sys/utsname.h>
+#include <stdio.h>
+#include <string.h>
```
```suggestion
#include <stdio.h>
#include <string.h>

#include <zephyr/posix/sys/utsname.h>
```

cfriedt commented on lib/posix/utsname.c:
```diff
@@ -0,0 +1,24 @@
+#include <stdio.h>
+#include <string.h>
+
+int uname(struct utsname *name)
```
I'm not usually one for adding to the ROM footprint, but it might be significantly simpler to have 1 `static const struct utsname` in this file and then `memcpy()` it to the destination.

It makes sense why we have the array sizes now, so please disregard my earlier Q.

ycsin replied:
do we need a mechanism to detect if hostname is changed if `net_hostname_set_postfix` is enabled, and update that static copy accordingly?

cfriedt replied:
Ok then, `memcpy()` + `snprintf()` for the hostname.

cfriedt reviewed (changes requested):
LGTM - just a couple of nits and suggestions.

The sample is good, but it would also be good to add to the testsuite.

One tricky area in particular is the version string.

We actually do want to be fairly consistent here with e.g. the boot banner.

So it would make sense to detect and modify the string such that we can display
* a tagged release (e.g. `v3.5.0`)
* something commit-ish (IIRC git short ref plus epoch time? I can't remember)
* whether the build is dirty (Linux uses a `+`, not sure how Zephyr does it)

Of possible, can you please also update `doc/services/portability/posix.rst`?

Great work! Thanks @ycsin 🚀🪁👍🍍

cfriedt commented:
Some compliance fixes..

cfriedt commented:
IIRC, we cannot yet build for `native_posix`. I would suggest filtering `CONFOG_ARCH_POSIX` in your sample

ycsin commented:
> We actually do want to be fairly consistent here with e.g. the boot banner.

The `version` string is the same as the [boot banner](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/banner.c#L19) (I just copied & renamed)

> * a tagged release (e.g. `v3.5.0`)
> * something commit-ish (IIRC git short ref plus epoch time? I can't remember)

Should be covered by:
```
3.4.99 zephyr-v3.4.0-632-gfed4ab1928ba Jul  4 2023 19:01:24
```

Not epoch time though..

> whether the build is dirty (Linux uses a `+`, not sure how Zephyr does it)

Looks like something to be done in the build system, not sure if @tejlmand has any idea?
I vaguely rmb that we used to have this somewhere..?

ycsin committed:
samples: posix: add sample for uname

Add a simple, testable sample to print everything in the utsname struct,
and added a `uname` shell cmd basically copied from nuttx just for fun.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: posix: uname: add test

add ztest for uname api

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on lib/posix/utsname.c:
```diff
@@ -0,0 +1,31 @@
+{
+	if (name == NULL) {
+		return -1;
+		errno = EINVAL;
```
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html:

> Upon successful completion, a non-negative value shall be returned. Otherwise, -1 shall be returned and errno set to indicate the error.

cfriedt replied:
I wouldn't even bother with the runtime check.  "No errors defined".
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

ycsin replied:
Replaced that runtime check with `__ASSERT_NO_MSG`

cfriedt commented on lib/posix/utsname.c:
```diff
@@ -0,0 +1,31 @@
+
+int uname(struct utsname *name)
+{
+	if (name == NULL) {
```
I wouldn't even bother with the runtime check.  "No errors defined".
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

tejlmand commented:
> > whether the build is dirty (Linux uses a `+`, not sure how Zephyr does it)
> 
> Looks like something to be done in the build system, not sure if @tejlmand has any idea? I vaguely rmb that we used to have this somewhere..?

Adding `-dirty` (or anything indicating a dirty build) has been rejected in the past, ref: https://github.com/zephyrproject-rtos/zephyr/pull/13608#pullrequestreview-206725920

But the use-case / reason for not having `-dirty` can be handled manually by those who has special needs, as described here:
https://github.com/zephyrproject-rtos/zephyr/issues/39503#issuecomment-961686122

As people has various ways of working / verifying patches, then I think a wider audience should be invited to discuss if we want to have `-dirty` (or similar) appended to short sha.
The implementation itself is very trivial, so just a matter of making a decision.

cfriedt commented on lib/posix/utsname.c:
```diff
@@ -0,0 +1,29 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/posix/sys/utsname.h>
+
+int uname(struct utsname *name)
+{
+	__ASSERT_NO_MSG((name != NULL));
+
+	strncpy(name->sysname, "Zephyr", sizeof(name->sysname));
+	strncpy(name->nodename, net_hostname_get(), sizeof(name->nodename));
+	strncpy(name->release, KERNEL_VERSION_STRING, sizeof(name->release));
+#if defined(__DATE__) && defined(__TIME__)
+	snprintf(name->version, sizeof(name->version), "%s %s %s", VERSION_BUILD, __DATE__,
+		 __TIME__);
+#else
+	strncpy(name->version, VERSION_BUILD, sizeof(name->version));
+#endif
+	strncpy(name->machine, CONFIG_ARCH, sizeof(name->machine));
+
+	return 0;
+}
```
```suggestion
#include <string.h>

#include <zephyr/kernel.h>
#include <zephyr/posix/sys/utsname.h>

#ifdef CONFIG_NET_HOSTNAME_ENABLE
#define UTSNAME_NODENAME CONFIG_NET_HOSTNAME
#else
#define UTSNAME_NODENAME "zephyr"
#endif

#if defined(__DATE__) && defined(__TIME__)
#define UTSNAME_VERSION(_ver) _ver " " __DATE__ " " __TIME__
#else
#define UTSNAME_VERSION(_ver) _ver
#endif

#define UTSNAME_INITIALIZER(_sys, _node, _rel, _ver, _mach)                                        \
	{                                                                                          \
		.sysname = _sys, .nodename = _node, .release = _rel,                               \
		.version = UTSNAME_VERSION(_ver), .machine = _mach,                                \
	}

static const struct utsname z_name = UTSNAME_INITIALIZER(
	"Zephyr", UTSNAME_NODENAME, KERNEL_VERSION_STRING, VERSION_BUILD, CONFIG_ARCH);

int uname(struct utsname *name)
{
	memcpy(name, &z_name, sizeof(*name));
	if (IS_ENABLED(CONFIG_NET_HOSTNAME_ENABLE)) {
		strncpy(name->nodename, net_hostname_get(), sizeof(name->nodename));
	}

	return 0;
}
```

cfriedt commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,84 @@
+#ifndef ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+#define ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+
+#include "version.h"
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <zephyr/net/hostname.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/net/hostname.h>
```
```suggestion
#include <zephyr/kernel.h>
```

ycsin replied:
moved all #include into the source file

cfriedt commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,84 @@
+
+/* sysname */
+#define SYS_NAMELEN sizeof("Zephyr")
+
+/* nodename */
+#ifdef CONFIG_NET_HOSTNAME_ENABLE
+#define NODE_NAMELEN NET_HOSTNAME_MAX_LEN
+#else
+#define NODE_NAMELEN SYS_NAMELEN
+#endif
+
+/* release */
+#define RELEASE_NAMELEN sizeof(KERNEL_VERSION_STRING)
+
+/* version */
+#if defined(__DATE__) && defined(__TIME__)
+#define DATETIME_LEN (sizeof(__DATE__) + sizeof(__TIME__))
+#else
+#define DATETIME_LEN 0
+#endif
+
+#ifdef BUILD_VERSION
+#define VERSION_BUILD STRINGIFY(BUILD_VERSION)
+#else
+#define VERSION_BUILD KERNEL_VERSION_STRING
+#endif
+#define VERSION_BUILD_LEN sizeof(VERSION_BUILD)
+
+/**
+ * In case anyone wondered if this length has been adjusted for '\0', it will
+ * be formatted in utsname.c as:
+ *     "%s %s %s", VERSION_BUILD, __DATE__, __TIME__
+ * the '\0' of VERSION_BUILD_LEN & __DATE__ will be used for the space that
+ * follows, '\0' of __TIME__ will be used for the \0 of the formatted VERSION
+ * string. In the end the size of the array should fit the string snugly.
+ */
+#define VERSION_NAMELEN (VERSION_BUILD_LEN + DATETIME_LEN)
+
+/* machine */
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
```
While I appreciate that this is all very precise, I kind of think we would be better off keeping it simple for now.

E.g. add a Kconfig integer option `CONFIG_POSIX_UTSNAME_LENGTH`, and use
```
#ifdef CONFIG_POSIX_UTSNAME
#define _UTSNAME_LENGTH CONFIG_POSIX_UTSNAME_LENGTH
#else
#define _UTSNAME_LENGTH sizeof("Zephyr")
#endif
```

https://man7.org/linux/man-pages/man2/uname.2.html

Inside of the `utsname.c` (which may be better to name `uname.c`), we can have several compile-time `BUILD_ASSERT()` statements. E.g.

```cpp
BUILD_ASSERT(sizeof(z_name.sysname) >= sizeof("Zephyr"));
BUILD_ASSERT(sizeof(z_name.release) >= sizeof(KERNEL_VERSION_STRING));
BUILD_ASSERT(sizeof(z_name.version) >= sizeof(UTSNAME_VERSION(VERSION_BUILD));
BUILD_ASSERT(sizeof(z_name.machine) >= sizeof(CONFIG_ARCH));

```

ycsin replied:
@keith-packard if we remove the `CONFIG_POSIX_UNAME_VERSION_LEN`, then we will have to reinstate the complicated part of the computation here

cfriedt commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,84 @@
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
+
+struct utsname {
+	char sysname[SYS_NAMELEN];
+	char nodename[NODE_NAMELEN];
+	char release[RELEASE_NAMELEN];
+	char version[VERSION_NAMELEN];
+	char machine[MACHINE_NAMELEN];
```
```suggestion
	char sysname[_UTSNAME_LENGTH];
	char nodename[_UTSNAME_LENGTH];
	char release[_UTSNAME_LENGTH];
	char version[_UTSNAME_LENGTH];
	char machine[_UTSNAME_LENGTH];
```

cfriedt commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,84 @@
+
+int uname(struct utsname *name);
+
+#undef SYS_NAMELEN
+#undef NODE_NAMELEN
+#undef RELEASE_NAMELEN
+#undef VERSION_NAMELEN
+#undef MACHINE_NAMELEN
+#undef DATETIME_LEN
+#undef VERSION_BUILD_LEN
+
```
```suggestion
```

cfriedt commented on lib/posix/Kconfig:
```diff
@@ -182,3 +182,10 @@ config FNMATCH
+	default y if POSIX_API
+	help
+	  The uname() function shall store information identifying the current
+	  system in the structure pointed to by name.
```
Would be good to also add `config POSIX_UTSNAME_LENGTH`, see above

cfriedt commented on samples/posix/uname/src/main.c:
```diff
@@ -0,0 +1,186 @@
+/*
```
This is great - we may actually want to move this into `lib/posix` at some point. Thanks for adding it :-)

cfriedt commented:
An alternative internal thing for `uname.c` is below, where we don't copy constant string data into an array.

With this setup, it would make sense to use `strncpy()` for each field.
```cpp
struct utsname_internal {
	const char *const sysname;
	const char *const nodename;
	const char *const release;
	const char *const version;
	const char *const machine;
};
```

cfriedt commented:
> Not epoch time though..

`__TIME_UNIX__` might work - I think it's epoch time

keith-packard commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -0,0 +1,27 @@
+	char sysname[sizeof("Zephyr")];
+	char nodename[CONFIG_POSIX_UNAME_NODENAME_LEN];
+	char release[sizeof("99.99.99")];
+	char version[CONFIG_POSIX_UNAME_VERSION_LEN];
```
I think the size of this field can also be computed -- just need to insert the computation of the version string into this header. One less config value to deal with seems like a nice thing.

ycsin replied:
We can actually compute everything using preprocessor defines here, at the expense of including 2 more headers & some macro magic then looks a bit complicated

cfriedt replied:
The version field can be fixed size as well, technically, too, if we omit the date and time, and it's compatible with the banner version as well.

```cpp
char version[sizeof("zephyr-v99.99.99-9999-g0123456789ab")];
```

Ideally we could just include `<version.h>`, which is generated at build time and use the exact `#define`s that appear there.

I'm hesitant to do that, just because it really isn't referenced outside of `banner.c`, which gives me the impression that it's private somehow.

If that makes life easier though, by all means, just 
```cpp
struct utsname {
	char sysname[sizeof("Zephyr")];
	char nodename[CONFIG_POSIX_UNAME_NODENAME_LEN + 1];
	char release[sizeof(KERNEL_VERSION_STRING)];
	char version[sizeof(STRINGIFY(BUILD_VERSION))];
	char machine[sizeof(CONFIG_ARCH)];
};
```

And honestly, the `__DATE__` and `__TIME__` macros are probably not *super* critical.

marc-hb replied:
`__DATE__` and `__TIME__` break build reproducibility and should not be enabled by default (not sure what is the current situation)

keith-packard commented on lib/posix/Kconfig:
```diff
@@ -182,3 +182,26 @@ config FNMATCH
+	help
+	  Size of the version char array in the utsname struct.
+
+config POSIX_UNAME_NODENAME_LEN
```
I looked at the Linux version of this code and it defines the maximum string length and then adds one when declaring the array to account for the null terminator. Seems like a good idea. Would mean changing the doc strings here to talk about 'maximum length of the node name'. (the doc string here was also copied from the version one above).

keith-packard commented on lib/posix/uname.c:
```diff
@@ -0,0 +1,52 @@
+{
+	memcpy(name, &z_name, sizeof(*name));
+	if (IS_ENABLED(CONFIG_NET_HOSTNAME_ENABLE)) {
+		strncpy(name->nodename, net_hostname_get(), sizeof(name->nodename));
```
strncpy doesn't ensure the result is null terminated.

ycsin replied:
added `\0` to the end of the buffer

keith-packard reviewed (changes requested):
the strncpy usage needs to be fixed, otherwise my comments are just suggestions.

ycsin commented:
> Adding `-dirty` (or anything indicating a dirty build) has been rejected in the past

how about having a separate define to indicate this in "version.h", rather than appending to the hash directly?

tejlmand commented:
> how about having a separate define to indicate this in "version.h", rather than appending to the hash directly?

Did you read the links ?
Especially: https://github.com/zephyrproject-rtos/zephyr/pull/13608#pullrequestreview-206725920 .

Even if placing this info in a separate define to use in `version.h` then it will impact the hex file when the repo goes from clean --> dirty (or other way around).

So if this is to be changed, then please open a dedicated issue for discussing, and not in this PR.

cfriedt commented on samples/posix/uname/src/main.c:
```diff
@@ -0,0 +1,186 @@
+	/* Process each option */
+
+	first = true;
+	for (i = 0; set != 0; i++) {
```
This can be done without a for loop so that it uses symbolic values instead of magic numbers.

e.g. `if (set & UNAME_VERSION)`

cfriedt reviewed (changes requested):
One minor change request

cfriedt commented:
> So if this is to be changed, then please open a dedicated issue for discussing, and not in this PR.

@tejlmand - I believe @ycsin's changes have addressed your concerns, and with a couple of small changes, I'm ready to approve. Can you take a look again to ensure that your concerns were addressed?

cfriedt commented on samples/posix/uname/src/main.c:
```diff
@@ -0,0 +1,186 @@
+	return 0;
+}
+
+#define UNAME_KERNEL   (1 << 0)
```
@ycsin - sorry - I missed this (was reviewing on mobile).

Please use `BIT(n)` here

cfriedt commented on tests/posix/common/src/uname.c:
```diff
@@ -0,0 +1,17 @@
+#include <zephyr/ztest.h>
+#include <sys/utsname.h>
+
+ZTEST(posix_apis, test_posix_uname)
```
@ycsin - one last thing (sorry again). Let's shorten this to `test_uname`).

cfriedt commented:
@keith-packard - does this one look OK to you now?

cfriedt commented:
@ycsin - I'm OK getting 1 more approval for this and merging if you wouldn't mind a follow-up PR for namespacing `<version.h>` as `<zephyr/kernel/banner_version.h>` or something along those lines.

keith-packard reviewed (approved):
Yeah, given the concerns about mixing too many bits into the application namespace, having compile-time checks that make sure the sizes are big enough seems good enough here.
